srcflib/tasks/membership.py

```diff
@@ -3,192 +3,291 @@
-from ..plumbing.mysql import context as mysql_context
-from . import mysql, pgsql
+from ..plumbing import bespoke, pgsql as pgsql_p, unix
+from ..plumbing.common import Collect, Password, Result, State
+from . import mailman, mysql, pgsql
+
+
+# Re-export enums to avoid unnecessary imports elsewhere.
+MailHandler = bespoke.MailHandler
 
 
+@Result.collect
 def create_member(crsid: str, preferred_name: str, surname: str, email: str,
-                  mail_handler: str, is_member: bool = True, is_user: bool = True,
-                  social: bool = False) -> ResultSet[Tuple[Member, Password]]:
+                  mail_handler: MailHandler, is_member: bool = True,
+                  is_user: bool = True, social: bool = False, 
+                  new_passwd: bool = False) -> Collect[Tuple[Member, Optional[Password]]]:
     """
     Register and provision a new member of the SRCF.
     """
-    results = ResultSet[Tuple[Member, Password]]()
     with bespoke.context() as sess:
-        result = bespoke.create_member(sess, crsid, preferred_name, surname, email,
-                                       mail_handler, is_member, is_user)
-        mem = results.add(result).value
-    user = results.add(unix.create_user(crsid, uid=mem.uid, system=True,
-                                        home_dir=os.path.join("/home", crsid),
-                                        real_name=mem.name)).value
-    passwd = results.add(unix.reset_password(user)).value
-    results.extend(bespoke.update_nis(True),
-                   unix.create_home(user, os.path.join("/public/home", crsid), True),
-                   bespoke.set_home_exim_acl(mem),
-                   bespoke.populate_home_dir(mem),
-                   bespoke.update_quotas())
+        res_record = yield from bespoke.ensure_member(sess, crsid, preferred_name, surname, email,
+                                                      mail_handler, is_member, is_user)
+        member = res_record.value
+    res_user = yield from unix.ensure_user(crsid, uid=member.uid, system=True,
+                                           home_dir=os.path.join("/home", crsid),
+                                           real_name=member.name)
+    new_user = res_user.state == State.created
+    user = res_user.value
+    if new_user or new_passwd:
+        res_passwd = yield from unix.reset_password(user)
+        passwd = res_passwd.value
+    else:
+        passwd = None
+    if res_user or passwd:
+        yield bespoke.update_nis(new_user)
+    yield unix.create_home(user, os.path.join("/public/home", crsid), True)
+    yield bespoke.set_home_exim_acl(member)
+    yield bespoke.populate_home_dir(member)
+    if res_record:
+        yield bespoke.update_quotas()
     if mail_handler == "pip":
-        results.extend(bespoke.create_forwarding_file(mem))
+        yield bespoke.create_forwarding_file(member)
     # TODO: Legacy mailbox creation
```

Comment:
I think legacy mailbox creation is as simple as sending an email to `real-$CRSID@srcf.net` (at least that's what we do currently)

Reply:
Indeed, I have this done on another branch already (where `srcflib.email` is a bit more buffed) -- I thought it best not to completely overload this PR!